libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -955,23 +991,42 @@ std::string IRGenerator::deployCode(ContractDefinition const& _contract)
 	if (_contract.isLibrary())
 	{
 		solAssert(ContractType(_contract).immutableVariables().empty(), "");
-		immutables.emplace_back(std::map<std::string, std::string>{
-			{"immutableName"s, IRNames::libraryAddressImmutable()},
-			{"value"s, "address()"}
-		});
-
+		if (!eof)
+			immutables.emplace_back(std::map<std::string, std::string>{
+				{"immutableName"s, IRNames::libraryAddressImmutable()},
+				{"value"s, "address()"}
+			});
+		else
+		{
+			auto const& immutableVariables = ContractType(_contract).immutableVariables();
+			auto const varIt = std::find_if(immutableVariables.begin(), immutableVariables.end(), [](VariableDeclaration const* _varDecl) { return _varDecl->name() == IRNames::libraryAddressImmutable(); });
+			solAssert(varIt != immutableVariables.end(), "");
+			immutables.emplace_back(std::map<std::string, std::string>{
+				{"immutablesMemoryStart"s, std::to_string(CompilerUtils::generalPurposeMemoryStart)},
+				{"immutablesOffset"s, std::to_string(m_context.immutableMemoryOffset(*(*varIt)))},
+				{"value"s, "address()"}
+			});
+		}
 	}
 	else
 		for (VariableDeclaration const* immutable: ContractType(_contract).immutableVariables())
 		{
 			solUnimplementedAssert(immutable->type()->isValueType());
 			solUnimplementedAssert(immutable->type()->sizeOnStack() == 1);
-			immutables.emplace_back(std::map<std::string, std::string>{
-				{"immutableName"s, std::to_string(immutable->id())},
-				{"value"s, "mload(" + std::to_string(m_context.immutableMemoryOffset(*immutable)) + ")"}
-			});
+			if (!eof)
+				immutables.emplace_back(std::map<std::string, std::string>{
+					{"immutableName"s, std::to_string(immutable->id())},
+					{"value"s, "mload(" + std::to_string(m_context.immutableMemoryOffset(*immutable)) + ")"}
+				});
+			else
+				immutables.emplace_back(std::map<std::string, std::string>{
+					{"immutablesMemoryStart"s, std::to_string(CompilerUtils::generalPurposeMemoryStart)},
+					{"immutablesOffset"s, std::to_string(m_context.immutableMemoryOffset(*immutable))},
+					{"value"s, "address()"}
```

Comment:
Legacy for some reason makes this the same way. I assume for library the address it not stored in `reservedMemory`.

Reply:
But this is the code for the non-library case. Or actually was - because I see it has been changed now.

Reply:
> I assume for library the address it not stored in `reservedMemory`.

It's not. It could have been, but I for legacy not having to allocate memory for it is actually better. On EOF on the other hand we need the memory anyway, so I'd store it there (https://github.com/ethereum/solidity/pull/15512#discussion_r1811634321).